scripts/collate_fastq_reads.R

```diff
@@ -30,40 +36,60 @@ if (args$out == "") {
+# Parse some parameters into vectors ----
 sequencing_index_cols= unlist(strsplit(args$sequencing_index_cols, ","))
 id_cols= unlist(strsplit(args$id_cols, ","))
 
-# Validation: Check that sequencing_index_cols are from sample meta column names
-if(!all(sequencing_index_cols %in% colnames(sample_meta))) {
-  stop(paste('The following sequencing_index_cols were not found in the sample meta: ',
-             sequencing_index_cols[!sequencing_index_cols %in% colnames(sample_meta)]))
+# Validation: Check that sequencing_index_cols present in the sample meta ----
+if(!validate_columns_exist(sequencing_index_cols, sample_meta)) {
+  stop('One or more sequencing_index_cols is NOT present in the sample meta.')
 }
 
-# Validation: Check that id_cols are from sample meta column names
-if(!all(id_cols %in% colnames(sample_meta))) {
-  stop(paste('The following id_cols were not found in the sample meta: ',
-             id_cols[!id_cols %in% colnames(sample_meta)]))
+# Validation: Check that id_cols are present in the sample meta ----
+if(!validate_columns_exist(id_cols, sample_meta)) {
+  stop('One or more id_cols is NOT present in the sample meta.')
 }
 
-print("Calling collate_fastq_reads ...")
-raw_counts= collate_fastq_reads(uncollapsed_raw_counts= raw_counts_uncollapsed, 
-                                sample_meta= sample_meta, 
-                                sequencing_index_cols= sequencing_index_cols,
-                                id_cols= id_cols,
-                                reverse_index2= args$reverse_index2,
-                                barcode_col= args$barcode_col)
+# Run collate_fastq_reads on chunks of raw_counts_uncollapsed.csv ----
+# raw_counts_uncollapsed can be too large to read into memory,
+# so collate_fastq_reads is performed on chunks of the large file.
+chunked_results= process_in_chunks(large_file_path= args$raw_counts_uncollapsed, 
+                                   chunk_size= 10^6, 
+                                   action= collate_fastq_reads,
+                                   # Parameters for collate_fastq_reads
+                                   sample_meta= sample_meta, 
+                                   sequencing_index_cols= sequencing_index_cols,
+                                   id_cols= id_cols,
+                                   known_barcodes= unique(c(cell_line_meta$Sequence, CB_meta$Sequence)),
+                                   reverse_index2= args$reverse_index2,
+                                   barcode_col= args$barcode_col,
+                                   low_abundance_threshold= as.numeric(args$low_abundance_threshold))
 
-# Validation: Basic file size check
-if(nrow(raw_counts) == 0) {
-  stop('ERROR: Empty file generated. No rows in raw_counts output.')
-} 
+# From each chunk, extract prism_barcode_counts and bind the rows together into one dataframe.
+prism_barcode_counts= data.table::rbindlist(lapply(chunked_results, function(x) x$prism_barcode_counts))
```

Comment:
Question about edge case:
will this lapply work smoothly if the list chunked_results has length of 1 or 0?
(corresponding to a file with either nrows < chunk_size and nrows=0)